File: solutions/innerprod_red.f90
PROGRAM innerprod

USE omp_lib
IMPLICIT NONE

INTEGER, PARAMETER :: N = 10000000
INTEGER, ALLOCATABLE :: A(:), B(:)
INTEGER :: ip, i
DOUBLE PRECISION :: t1, t2

    ALLOCATE(A(N),B(N))
    DO i = 1, N
       A(i) = 2
       B(i) = 3
    ENDDO

    t1 = OMP_GET_WTIME()
    ip = 0
!$OMP PARALLEL DO REDUCTION(+:ip)
    DO i = 1, N
       ip = ip + A(i) * B(i)
    ENDDO
!$OMP END PARALLEL DO
    t2 = OMP_GET_WTIME()

    WRITE(*,*) 'Inner product = ', ip, 'Time =', t2 - t1

END PROGRAM innerprod

File: solutions/innerprod_red.c
/* innerprod.c */
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

#define N 10000000

int main(void)
{
    int *a, *b, ip, i;
    double t1, t2;

    a = malloc(N * sizeof(*a));
    b = malloc(N * sizeof(*b));

    for (i = 0; i < N; i++) {
        a[i] = 2;
        b[i] = 3;
    }

    t1 = omp_get_wtime();
    ip = 0;

#pragma omp parallel for private(i) shared (a,b) reduction(+:ip)
      for (i = 0; i < N; i++)
        ip += a[i] * b[i];

    t2 = omp_get_wtime();

    printf("Inner product = %d, time=%g\n", ip, t2-t1);
    return 0;
}
